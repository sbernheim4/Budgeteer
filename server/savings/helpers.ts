import { IBankInfo, IInstitutionSavingsPoint, IInstitutionSavingsInfo } from './../types';

/**
 * @description Converts IBankInfo[] to an of IInstitutionSavingsInfo[]
 * @param {IBankInfo} data The new data point to be converted
 * @returns {IInstitutionSavingsInfo} The normalized data
 */
export function formatNewDataPoints(data: IBankInfo): IInstitutionSavingsInfo {

	const dateString = new Date().toString();

	const { institutionId, institutionBalance } = data;

	const newInfo: IInstitutionSavingsInfo = {
		institutionId,
		savingsData: [{
			date: dateString,
			institutionalBalance: institutionBalance
		}]
	};

	return newInfo;

}

/**
 * @description Merges the existing data points with the new data points
 * @param {IInstitutionSavingsInfo[]} oldDataPoints An arrray containing the existing data points
 * @param {IInstitutionSavingsInfo[]} newDataPoints An array of new data points to be merged
 * @returns {IInstitutionSavingsInfo[]} The merged array of newDataPoints and savings
 */
export function createNewInstitutionData(oldDataPoints: IInstitutionSavingsInfo, newDataPoints: IInstitutionSavingsInfo): IInstitutionSavingsInfo {

	const { institutionId, savingsData } = newDataPoints;

	let newDataPointsArray: IInstitutionSavingsPoint[] = [];

	try {

		const oldInstitutionSavingsData = oldDataPoints.savingsData;

		oldInstitutionSavingsData.sort((a, b) => {
			return new Date(a.date).getTime() - new Date(b.date).getTime();
		});

		newDataPointsArray = [...oldInstitutionSavingsData, ...savingsData];

	} catch (error) {

		newDataPointsArray = savingsData;

	}

	const updatedInstitutionInfo: IInstitutionSavingsInfo = {
		institutionId,
		savingsData: newDataPointsArray
	}


	return updatedInstitutionInfo;

}

