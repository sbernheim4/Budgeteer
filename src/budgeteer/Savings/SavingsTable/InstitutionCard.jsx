import React, { useState, useRef, useEffect } from 'react';

import InstitutionInfoRow from './InstitutionInfoRow.jsx';
import SavingsChart from './../SavingsChart/SavingsChart.jsx';

export default function InstitutionCard(props) {

	const {
		institutionInfo,
		institutionNames,
		institutionId,
		totalSavings,
		storeNewChartData
	} = props;

	const [width, setWidth] = useState(0);
	const institutionNameRef = useRef(null);

	useEffect(() => {

		const width = institutionNameRef.current ? institutionNameRef.current.offsetWidth : 0;
		setWidth(width);

	}, [institutionNameRef.current]);

	function generateInstitutionInfoRows(institutionInfo) {

		const institutionInfoKeys = Object.keys(institutionInfo);

		const institutionInfoRows = institutionInfoKeys.map((acctId, i) => {
			return <InstitutionInfoRow
				key={i}
				institutionInfo={institutionInfo}
				totalSavings={totalSavings}
				acctId={acctId}
			/>
		});

		institutionInfoRows.sort((a, b) => {
			const aAccountId = a.props.acctId;
			const bAccountId = b.props.acctId;
			const aTotal = a.props.institutionInfo[aAccountId].accountBalance;
			const bTotal = b.props.institutionInfo[bAccountId].accountBalance;

			return bTotal - aTotal;
		});

		return institutionInfoRows;

	}

	const institutionInfoValues = Object.values(institutionInfo);
	const institutionBalanceTotal = institutionInfoValues.reduce((acc, currVal) => acc + currVal.accountBalance, 0);
	const institutionName = institutionNames[institutionId];
	const percent = Math.round(institutionBalanceTotal / totalSavings * 100, 2);
	const institutionInfoRows = generateInstitutionInfoRows(institutionInfo);

	function handleClick(e) {
		const card = e.target.closest(".institution-card").querySelector(".savings-chart");
		card.classList.toggle('savings-chart--show');
	}

	return (
		<div className='institution-card'>

			<div className='institution-card--info'
				onClick={(e) => handleClick(e)}
			>
				<h3 style={{left: `calc(50% - ${width / 2}px`}}>
					<span
						className='institution-card--info--name'
						ref={institutionNameRef}
					>
					{institutionName}
					</span>

					<span
						className='percent'
					> - {percent}%
					</span>

				</h3>

				{institutionInfoRows}
			</div>

			<SavingsChart
				institutionId={institutionId}
				storeNewChartData={storeNewChartData}
			/>
		</div>
	);
}
