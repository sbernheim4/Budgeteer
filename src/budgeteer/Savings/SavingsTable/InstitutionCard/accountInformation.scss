@import './../../../scss/variables.scss';

.institution-card--info--name--loading {
	display: inline-block;
	padding: 3px 15px;
	background-color: #dedede;
	border-radius: 10px;
	font-weight: bolder;
	color: #dedede;
	vertical-align: bottom;
}

.institution-card {
	padding: 20px;

	&--info {
		position: relative;
		margin: 0 auto;
		max-width: 600px;
		width: 80%;
		padding: 20px 30px;
		background: white;
		border-radius: 7px;
		cursor: default;
		box-shadow: 0px 0px 20px 0px darken($blue, 20%);
		border-collapse: collapse; // needed to show border on the tr element
		z-index: 5;

		@media all and (max-width: 500px) {
			padding: 20px 30px;
		}

		h3 {
			position: relative;
			margin-bottom: 5px;
			font-size: 17px;
			color: #343434;

			&--name {
				display: inline-block;
				vertical-align: bottom;
				font-weight: bolder;
			}

		}

		&--percent {
			position: relative;
			display: none;
			left: -5px;
		}

		&:hover {

			.institution-card--info--percent {
				display: inline;
				left: 0;
				animation: slideRight .2s ease-in-out;
			}

		}

		.networth--entry {
			margin: 0 auto;
			width: 95%;
			height: 50px;
			font-size: 17px;
			text-align: center;
			border-bottom: 1px solid #a9a9a9;
			display: grid;
			justify-content: center;
			align-items: center;
			grid-template-columns: 4fr 130px .5fr;
			grid-template-rows: 1fr;

			@media all and (max-width: 600px) {
				grid-template-columns: 6fr 130px 1fr;
			}

			@media all and (max-width: 500px) {
				grid-template-columns: 1fr 100px 0fr;
				font-size: 18px;
			}

			@media all and (max-width: 400px) {
				grid-template-columns: 1fr 90px 0fr;
				font-size: 16px;
			}

			.acct-name {
				text-align: left;
			}

			.acct-value {
				font-weight: 400;
			}

			.acct-percentage {
				width: 100%;
				min-width: 50px;
				height: 100%;
				min-height: 50px;
				place-self: center;

				@media all and (max-width: 500px) {
					display: none;
				}
			}
		}

		.networth--entry:nth-last-child(1) {
			border-bottom: none;
		}

		.networth--total {

			p {
				font-weight: bolder;
			}

		}

	}

}
