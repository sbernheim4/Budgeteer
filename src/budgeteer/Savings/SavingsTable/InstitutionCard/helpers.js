 export function determineNewWidth(institutionNameRef) {
	const width = institutionNameRef.current ? institutionNameRef.current.offsetWidth : 0;
	return width;
}
