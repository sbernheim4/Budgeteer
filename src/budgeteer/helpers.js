const helper = {

	numberWithCommas(number) {
		return number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
	},

	formatAmount(amt) {
		return (Math.round(amt * 100) / 100).toFixed(2);
	},

	toTitleCase(str) {
		if (str === "" || str === null) return "";
		return str.toLowerCase().split(' ').map(function(word) {
			return word.replace(word[0], word[0].toUpperCase());
		}).join(' ');
	}
}

export default helper
