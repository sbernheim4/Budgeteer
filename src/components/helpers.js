const helper = {

	numberWithCommas(number) {
		return number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
	},

	formatAmount(amt) {
		return (Math.round(amt * 100) / 100).toFixed(2);
	}

}

export default helper
