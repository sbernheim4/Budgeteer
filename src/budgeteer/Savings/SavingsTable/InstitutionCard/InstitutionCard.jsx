import React, { useState, useRef, useEffect, useCallback } from 'react';
import AccountInformation from '../AccountInformation.jsx';
import SavingsChart from './../../SavingsChart/SavingsChart.jsx';
import { determineNewWidth } from './helpers';

import './accountInformation.scss';

export default function InstitutionCard(props) {

	const {
		institutionInfo,
		institutionNames,
		institutionId,
		totalSavings,
	} = props;

	const [width, setWidth] = useState(0);
	const [displayChart, setDisplayChart] = useState(false);

	const institutionNameRef = useRef(null);
	const positionBankName = useCallback(() => determineNewWidth(institutionNameRef), [props.institutionNames]);

	useEffect(() => {

		const newWidth = positionBankName(institutionNameRef);
		setWidth(newWidth);

	}, [props.institutionNames]);

	function displayInstitutionAccounts(institutionInfo, totalSavings) {

		const accountIds = Object.keys(institutionInfo);

		const institutionAccounts = accountIds.map((accountId, i) => {

			return <AccountInformation
				key={i}
				institutionInfo={institutionInfo}
				totalSavings={totalSavings}
				acctId={accountId}
			/>

		}).sort((a, b) => {

			const aAccountId = a.props.acctId;
			const bAccountId = b.props.acctId;

			const aTotal = a.props.institutionInfo[aAccountId].accountBalance;
			const bTotal = b.props.institutionInfo[bAccountId].accountBalance;

			return bTotal - aTotal;

		});

		return institutionAccounts;

	}

	const institutionInfoValues = Object.values(institutionInfo);
	const institutionBalanceTotal = institutionInfoValues.reduce(
		(acc, currVal) => acc + currVal.accountBalance,
		0
	);

	const percent = Math.round(institutionBalanceTotal / totalSavings * 100, 2) === Infinity ?
		'...' :
		Math.round(institutionBalanceTotal / totalSavings * 100, 2);

	const institutionName = institutionNames[institutionId] || 'loading';

	const instituitionClassName = institutionName === 'loading' ?
		'institution-card--info--name institution-card--info--name--loading' :
		'institution-card--info--name';

	return (
		<div className='institution-card'>

			<div className='institution-card--info'
				onClick={() => { setDisplayChart(!displayChart); }}
			>
				<h3 style={{left: `calc(50% - ${width / 2}px`}}>

					<span
						className={instituitionClassName}
						ref={institutionNameRef}
					>
						{institutionName}
					</span>

					<span
						className='institution-card--info--percent'
					> - {percent}%
					</span>

				</h3>

				{displayInstitutionAccounts(institutionInfo, totalSavings)}

			</div>

			<SavingsChart
				displayChart={displayChart}
				institutionId={institutionId}
			/>
		</div>
	);
}
