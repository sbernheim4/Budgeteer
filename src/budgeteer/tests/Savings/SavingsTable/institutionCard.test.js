import React, { useRef } from 'react';
import { Provider } from 'react-redux';
import store from './../../../../../__mocks__/mockReduxStore';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import InstitutionCard from './../../../../budgeteer/Savings/SavingsTable/InstitutionCard/InstitutionCard';
import { determineNewWidth } from './../../../../budgeteer/Savings/SavingsTable/InstitutionCard/helpers';

configure({ adapter: new Adapter() });

describe('InstitutionCard Component', () => {


	const props = {
		institutionInfo: {
			accountId: 'accountId'
		},
		institutionNames: new Map().set('1', 'My Custom Bank Name'),
		institutionId: '1'
	}

	let wrapper = mount(<Provider store={store}> <InstitutionCard {...props}/> </Provider>);

	describe('component', () => {

		it('should render the same as always', () => {

			expect(wrapper).toMatchSnapshot();

		});

		it('should apply toggle the css class when clicked, and remove it when clicked again', () => {

			expect(wrapper.html().includes('show')).toBe(false);
			wrapper.find('div.institution-card--info').simulate('click');
			expect(wrapper.html().includes('savings-chart--show')).toBe(true);

			wrapper.find('div.institution-card--info').simulate('click');
			expect(wrapper.html().includes('savings-chart--show')).toBe(false);

		});

	});

	describe('helpers', () => {

		describe('#determineNewWidth', () => {

			const info = wrapper.find('.institution-card--info--name');
			const newWidth = determineNewWidth(info);
			expect(newWidth).toBe(0);

		});

	});

});

