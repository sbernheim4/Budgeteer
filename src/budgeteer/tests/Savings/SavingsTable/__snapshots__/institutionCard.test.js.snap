// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstitutionCard Component component should render the same as always 1`] = `ReactWrapper {}`;
